### ISSUE TITLE
mgmt/mcumgr/lib: Rework of event callback framework

### ISSUE THREAD

**Issue by de-nordic** [zephyrproject-rtos/zephyr#47274]
**Is your enhancement proposal related to a problem? Please describe.**
Currently mcumgr provides simplified callback subsystem that allows to register "event  manager" callback that will be called at some predefined points: before calling command handler, after return from handler and in image upload handler.
The event manager is registered with `mgmt_register_evt_cb` and within code is invoked by `mgmt_evt` function.
There are several issues with it:
 1) it is always compiled in, even if not used and does not allow to select which events will be compiled in;
 2) it does not allow to chain link handlers;
 3) It only covers handful of events;
 4) it has not been designed to allow stopping further execution by handler;
 5) the `MGMT_EVT_OP_CMD_DONE` is used for two execution paths, fortunately it is possible to distinguish between successful and unsuccessful by `mgmt_evt`  `arg` parameter.

The 3) has been fixed by addition of another set of callbacks: `img_mgmt_dfu_stopped` , `img_mgmt_dfu_started`, `img_mgmt_dfu_pending`, `img_mgmt_dfu_confirmed` with register function `img_mgmt_register_callbacks`;
which have following issues:
 1) it is always compiled in, even if not used and does not allow to select which events will be compiled in;
 2) it does not allow to chain link handlers;
 3) it only covers handful of events and exists in img management only;
 4) it has not been designed to allow stopping further execution by handler;
 5) the `img_mgmt_dfu_stopped` is called from two different paths and does not allow to figure out where it has been called from.

So basically we have two sets of events/handlers/callbacks that somehow work and have the same problems, just cover different areas in code.

**Describe the solution you'd like**
One solid, configurable and possible to complete disable at compile time, event framework that would allow to not only record an event but also take action that would affect further command processing.

**Describe alternatives you've considered**
Is there alternative to not fixing this?

**Additional context**
TBD
